File: javatest/omnia/data/cache/SimpleCacherTest.java
package omnia.data.cache;

import static org.junit.Assert.assertSame;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.function.Supplier;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class SimpleCacherTest {

  @Test
  public void value_didReturnSuppliedValue() {
    Object testValue = new Object();
    Cached<Object> testSubject = new SimpleCacher<>(() -> testValue);

    assertSame(testValue, testSubject.value());
  }

  @Test
  public void value_twice_didReturnSuppliedValue() {
    Object testValue = new Object();
    Cached<Object> testSubject = new SimpleCacher<>(() -> testValue);

    testSubject.value();

    assertSame(testValue, testSubject.value());
  }

  @Test
  public void new_didNotInvokeSupplier() {
    Supplier<Object> supplier = setUpMockSupplier();
    new SimpleCacher<>(supplier);

    verify(supplier, never()).get();
  }

  @Test
  public void invalidate_didNotSupplier() {
    Supplier<Object> supplier = setUpMockSupplier();
    Cached<Object> testSubject = new SimpleCacher<>(supplier);

    testSubject.invalidate();

    verify(supplier, never()).get();
  }

  @Test
  public void value_didInvokeSupplier() {
    Supplier<Object> supplier = setUpMockSupplier();
    Cached<Object> testSubject = new SimpleCacher<>(supplier);

    testSubject.value();

    verify(supplier).get();
  }

  @Test
  public void value_didInvokeSupplierOnce() {
    Supplier<Object> supplier = setUpMockSupplier();
    Cached<Object> testSubject = new SimpleCacher<>(supplier);

    testSubject.value();
    testSubject.value();

    verify(supplier, times(1)).get();
  }

  @Test
  public void value_thenInvalidate_thenValue_didInvokeSupplierTwice() {
    Supplier<Object> supplier = setUpMockSupplier();
    Cached<Object> testSubject = new SimpleCacher<>(supplier);

    testSubject.value();
    testSubject.invalidate();
    testSubject.value();

    verify(supplier, times(2)).get();
  }

  @Test(expected = NullPointerException.class)
  public void new_withNullSupplier_didThrowException() {
    new SimpleCacher<>(null);
  }

  @Test(expected = NullPointerException.class)
  public void value_whenSupplierReturnsNull_didThrowException() {
    Cached<Object> testSubject = new SimpleCacher<>(() -> null);

    testSubject.value();
  }

  private static Supplier<Object> setUpMockSupplier() {
    Supplier<Object> supplier = mock(ObjectSupplier.class);
    when(supplier.get()).thenReturn(new Object());
    return supplier;
  }

  interface ObjectSupplier extends Supplier<Object> {}
}

File: omnia/src/commonTest/omnia/data/cache/SimpleCacherTest.kt
package omnia.data.cache

import kotlin.test.Test
import omnia.util.test.fluent.Assertion.Companion.assertThat
import omnia.util.test.fluent.isEqualTo
import omnia.util.test.fluent.isSameAs

class SimpleCacherTest {

  @Test
  fun value_didReturnSuppliedValue() {
    val testValue = Any()
    val testSubject: Cached<Any> = SimpleCacher { testValue }
    assertThat(testSubject.value).isSameAs(testValue)
  }

  @Test
  fun value_twice_didReturnSuppliedValue() {
    val testValue = Any()
    val testSubject: Cached<Any> = SimpleCacher { testValue }
    testSubject.value
    assertThat(testSubject.value).isSameAs(testValue)
  }

  @Test
  fun new_didNotInvokeSupplier() {
    val supplier = setUpMockSupplier()
    SimpleCacher(supplier)
    assertThat(supplier.invocations).isEqualTo(0)
  }

  @Test
  fun invalidate_didNotInvokeSupplier() {
    val supplier = setUpMockSupplier()
    val testSubject = SimpleCacher(supplier)
    testSubject.invalidate()
    assertThat(supplier.invocations).isEqualTo(0)
  }

  @Test
  fun value_didInvokeSupplier() {
    val supplier = setUpMockSupplier()
    val testSubject = SimpleCacher(supplier)
    testSubject.value
    assertThat(supplier.invocations).isEqualTo(1)
  }

  @Test
  fun value_twice_didInvokeSupplierOnce() {
    val supplier = setUpMockSupplier()
    val testSubject = SimpleCacher(supplier)
    testSubject.value
    testSubject.value
    assertThat(supplier.invocations).isEqualTo(1)
  }

  @Test
  fun value_thenInvalidate_thenValue_didInvokeSupplierTwice() {
    val supplier = setUpMockSupplier()
    val testSubject = SimpleCacher(supplier)
    testSubject.value
    testSubject.invalidate()
    testSubject.value
    assertThat(supplier.invocations).isEqualTo(2)
  }

  companion object {
    private fun setUpMockSupplier() = MockSupplier(Any())
  }
}